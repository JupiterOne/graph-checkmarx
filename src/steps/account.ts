import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';
import { entities, ACCOUNT_ENTITY_DATA_KEY } from '../constants';

export function getAccountKey(name: string): string {
  return `checkmarx_account:${name}`;
}

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const company = {
    name: instance.config.instanceHostname,
  };

  const accountEntity = createIntegrationEntity({
    entityData: {
      source: company,
      assign: {
        _key: getAccountKey(company.name),
        _type: entities.ACCOUNT._type,
        _class: entities.ACCOUNT._class,
        webLink: `https://${company.name}.checkmarx.net`,
        name: company.name,
      },
    },
  });

  await Promise.all([
    jobState.addEntity(accountEntity),
    jobState.setData(ACCOUNT_ENTITY_DATA_KEY, accountEntity),
  ]);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
