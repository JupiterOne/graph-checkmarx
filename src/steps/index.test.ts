import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { fetchAccountDetails } from './account';
import { fetchServiceDetails } from './service';
import { fetchTeams } from './teams';
import { fetchProjects } from './projects';
import { fetchProjectScans } from './project-scans';
import { fetchScanFindings } from './scan-findings';
import { setupCheckmarxRecording } from '../../test/recording';

const DEFAULT_INSTANCE_HOSTNAME = 'partners9x';
const DEFAULT_CLIENT_USERNAME = 'INVALID';
const DEFAULT_CLIENT_PASSWORD = 'INVALID';

const integrationConfig: IntegrationConfig = {
  instanceHostname: process.env.INSTANCE_HOSTNAME || DEFAULT_INSTANCE_HOSTNAME,
  clientUsername: process.env.CLIENT_USERNAME || DEFAULT_CLIENT_USERNAME,
  clientPassword: process.env.CLIENT_PASSWORD || DEFAULT_CLIENT_PASSWORD,
};

jest.setTimeout(60 * 1000);

describe('Checkmarx', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupCheckmarxRecording({
      directory: __dirname,
      name: 'checkmarx_recordings',
      options: {
        recordFailedRequests: true,
      },
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    // Simulates dependency graph execution.
    // See https://github.com/JupiterOne/sdk/issues/262.
    await fetchAccountDetails(context);
    await fetchServiceDetails(context);
    await fetchTeams(context);
    await fetchProjects(context);
    await fetchProjectScans(context);
    await fetchScanFindings(context);

    // Review snapshot, failure is a regression
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities.map((e) => ({
        ...e,
        // HACK (austinkelleher): It looks like the formatting that Checkmarkx
        // uses does not include the timezone and it's causing issues with tests
        //
        // See a failing SDK test here: https://github.com/JupiterOne/sdk/commit/d88453dbf87e04802b6e81a398123b10e1a344f0#diff-eb9509d3689a5d1dde0aaf6a4f037d78ee0c178b48400ccc3249d8b84f726328R155
        createdOn: 123456789,
      })),
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Account'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_account' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          displayName: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Service'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Service'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_dast_scanner' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: {
            type: 'string',
          },
          displayName: {
            type: 'string',
          },
          category: {
            type: 'array',
          },
          function: {
            type: 'array',
            items: { type: 'string' },
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Team'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Team'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_team' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          id: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Project'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Project'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          id: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          isPublic: {
            type: 'boolean',
          },
          public: {
            type: 'boolean',
          },
          remoteSettingsLinkType: { type: 'string' },
          remoteSettingsUrl: { type: 'string' },
          remoteSettingsBranch: { type: 'string' },
          remoteSettingsUseSsh: { type: 'boolean' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Assessment'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Assessment'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_scan' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          id: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          category: {
            type: 'string',
          },
          project: {
            type: 'string',
          },
          summary: {
            type: 'string',
          },
          internal: {
            type: 'boolean',
          },
          status: {
            type: 'string',
          },
          isPublic: {
            type: 'boolean',
          },
          isLocked: {
            type: 'boolean',
          },
          scanRisk: {
            type: 'number',
          },
          scanRiskSeverity: {
            type: 'number',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Finding'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Finding'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'checkmarx_finding' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          id: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          category: {
            type: 'string',
          },
          severity: {
            type: 'string',
          },
          numericSeverity: {
            type: 'number',
          },
          open: {
            type: 'boolean',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });
  });
});
