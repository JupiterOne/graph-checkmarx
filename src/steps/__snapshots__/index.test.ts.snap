// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkmarx should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "checkmarx_account:cxprivatecloud",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "cxprivatecloud",
          },
        },
      ],
      "_type": "checkmarx_account",
      "createdOn": undefined,
      "displayName": "cxprivatecloud",
      "name": "cxprivatecloud",
      "webLink": "https://cxprivatecloud.checkmarx.net",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "checkmarx_dast_scanner:Checkmarx DAST",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "Checkmarx DAST",
          },
        },
      ],
      "_type": "checkmarx_dast_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "displayName": "Checkmarx DAST",
      "function": "DAST",
      "name": "Checkmarx DAST",
    },
    Object {
      "_class": Array [
        "Team",
      ],
      "_key": "checkmarx_team:109",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "fullName": "/CxServer/SP/Partners/JupiterOne",
            "id": 109,
            "name": "JupiterOne",
            "parentId": 0,
          },
        },
      ],
      "_type": "checkmarx_team",
      "createdOn": undefined,
      "displayName": "JupiterOne",
      "id": "109",
      "name": "JupiterOne",
    },
    Object {
      "_class": Array [
        "Project",
      ],
      "_key": "checkmarx_project:328",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "customFields": Array [],
            "id": 328,
            "isPublic": true,
            "name": "test-project",
            "teamId": 109,
          },
        },
      ],
      "_type": "checkmarx_project",
      "createdOn": undefined,
      "displayName": "test-project",
      "id": "328",
      "isPublic": true,
      "name": "test-project",
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/projectscans.aspx?id=328",
    },
    Object {
      "_class": Array [
        "Project",
      ],
      "_key": "checkmarx_project:352",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "customFields": Array [],
            "id": 352,
            "isPublic": true,
            "name": "another-project",
            "teamId": 109,
          },
        },
      ],
      "_type": "checkmarx_project",
      "createdOn": undefined,
      "displayName": "another-project",
      "id": "352",
      "isPublic": true,
      "name": "another-project",
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/projectscans.aspx?id=352",
    },
    Object {
      "_class": Array [
        "Project",
      ],
      "_key": "checkmarx_project:396",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "customFields": Array [],
            "id": 396,
            "isPublic": true,
            "name": "another-test-project",
            "teamId": 109,
          },
        },
      ],
      "_type": "checkmarx_project",
      "createdOn": undefined,
      "displayName": "another-test-project",
      "id": "396",
      "isPublic": true,
      "name": "another-test-project",
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/projectscans.aspx?id=396",
    },
    Object {
      "_class": Array [
        "Assessment",
      ],
      "_key": "checkmarx_scan:1000990",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "comment": "",
            "dateAndTime": Object {
              "engineFinishedOn": "2020-10-07T12:41:46.537",
              "engineStartedOn": "2020-10-07T12:41:21.443",
              "finishedOn": "2020-10-07T12:41:45.097",
              "startedOn": "2020-10-07T12:41:21.443",
            },
            "engineServer": Object {
              "id": 1,
              "link": null,
              "name": "Localhost",
            },
            "finishedScanStatus": Object {
              "id": 0,
              "value": "None",
            },
            "id": 1000990,
            "initiatorName": "Adam Williams",
            "isIncremental": false,
            "isLocked": false,
            "isPublic": true,
            "origin": "Web Portal",
            "owner": "adam.williams@jupiterone.com",
            "owningTeamId": 109,
            "partialScanReasons": null,
            "project": Object {
              "id": 328,
              "link": null,
              "name": "test-project",
            },
            "resultsStatistics": Object {
              "link": null,
            },
            "scanRisk": 10,
            "scanRiskSeverity": 0,
            "scanState": Object {
              "cxVersion": "9.2.0.41015 HF1",
              "failedLinesOfCode": 0,
              "filesCount": 1,
              "languageStateCollection": Array [
                Object {
                  "languageHash": "3450320970348195",
                  "languageID": 1073741824,
                  "languageName": "Common",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
                Object {
                  "languageHash": "0128891942401620",
                  "languageID": 8,
                  "languageName": "JavaScript",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
                Object {
                  "languageHash": "7713386434517252",
                  "languageID": 64,
                  "languageName": "VbScript",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
              ],
              "linesOfCode": 88,
              "path": " N/A (Zip File)",
              "sourceId": "0000000001_0-1072823060_0-1772354393",
            },
            "scanType": Object {
              "id": 1,
              "value": "Regular",
            },
            "status": Object {
              "details": Object {
                "stage": "",
                "step": "",
              },
              "id": 7,
              "name": "Finished",
            },
          },
        },
      ],
      "_type": "checkmarx_scan",
      "active": undefined,
      "category": "Vulnerability Scan",
      "createdOn": undefined,
      "displayName": "1000990",
      "id": "1000990",
      "internal": false,
      "isLocked": false,
      "isPublic": true,
      "name": "1000990",
      "owner": "adam.williams@jupiterone.com",
      "project": "test-project",
      "scanRisk": 10,
      "scanRiskSeverity": 0,
      "status": "Finished",
      "summary": "Regular",
    },
    Object {
      "_class": Array [
        "Assessment",
      ],
      "_key": "checkmarx_scan:1001371",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "comment": "",
            "dateAndTime": Object {
              "engineFinishedOn": "2020-11-10T16:02:50.883",
              "engineStartedOn": "2020-11-10T16:00:51.513",
              "finishedOn": "2020-11-10T16:02:47.85",
              "startedOn": "2020-11-10T16:00:51.513",
            },
            "engineServer": Object {
              "id": 1,
              "link": null,
              "name": "Localhost",
            },
            "finishedScanStatus": Object {
              "id": 0,
              "value": "None",
            },
            "id": 1001371,
            "initiatorName": "Adam Williams",
            "isIncremental": false,
            "isLocked": false,
            "isPublic": true,
            "origin": "Web Portal",
            "owner": "adam.williams@jupiterone.com",
            "owningTeamId": 109,
            "partialScanReasons": null,
            "project": Object {
              "id": 396,
              "link": null,
              "name": "another-test-project",
            },
            "resultsStatistics": Object {
              "link": null,
            },
            "scanRisk": 25,
            "scanRiskSeverity": 6,
            "scanState": Object {
              "cxVersion": "9.2.0.41015 HF4",
              "failedLinesOfCode": 162,
              "filesCount": 226,
              "languageStateCollection": Array [
                Object {
                  "languageHash": "3450320970348195",
                  "languageID": 1073741824,
                  "languageName": "Common",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
                Object {
                  "languageHash": "0128891942401620",
                  "languageID": 8,
                  "languageName": "JavaScript",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
                Object {
                  "languageHash": "0168023923119858",
                  "languageID": 1024,
                  "languageName": "Ruby",
                  "stateCreationDate": "2020-09-13T09:07:38.44",
                },
              ],
              "linesOfCode": 34080,
              "path": "https://github.com/JupiterOne/graph-azure",
              "sourceId": "0000000226_0-1951287486_000158255968",
            },
            "scanType": Object {
              "id": 1,
              "value": "Regular",
            },
            "status": Object {
              "details": Object {
                "stage": "",
                "step": "",
              },
              "id": 7,
              "name": "Finished",
            },
          },
        },
      ],
      "_type": "checkmarx_scan",
      "active": undefined,
      "category": "Vulnerability Scan",
      "createdOn": undefined,
      "displayName": "1001371",
      "id": "1001371",
      "internal": false,
      "isLocked": false,
      "isPublic": true,
      "name": "1001371",
      "owner": "adam.williams@jupiterone.com",
      "project": "another-test-project",
      "scanRisk": 25,
      "scanRiskSeverity": 6,
      "status": "Finished",
      "summary": "Regular",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "checkmarx_finding:projectid=328&scanid=1000990&pathid=1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "Assigned To": "",
            "Column": "1",
            "Comment": "",
            "Custom": "",
            "DestColumn": "1",
            "DestFileName": "rar/order.htm",
            "DestLine": "1",
            "DestName": "CxJSNS_44bcf57f",
            "DestNodeId": "1",
            "FISMA 2014": "",
            "Line": "1",
            "Link": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1000990&projectid=328&pathid=1",
            "NIST SP 800-53": "",
            "Name": "CxJSNS_44bcf57f",
            "NodeId": "1",
            "OWASP Mobile Top 10 2016": "",
            "OWASP Top 10 2013": "",
            "OWASP Top 10 2017": "",
            "PCI DSS v3.2": "",
            "Query": "Potential Clickjacking on Legacy Browsers",
            "QueryPath": "JavaScript\\\\Cx\\\\JavaScript Low Visibility\\\\Potential Clickjacking on Legacy Browsers Version:1",
            "Result Severity": "Low",
            "Result State": "To Verify",
            "Result Status": "New",
            "SrcFileName": "rar/order.htm",
          },
        },
      ],
      "_type": "checkmarx_finding",
      "category": "sast-scan",
      "createdOn": undefined,
      "displayName": "Potential Clickjacking on Legacy Browsers",
      "id": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1000990&projectid=328&pathid=1",
      "name": "Potential Clickjacking on Legacy Browsers",
      "numericSeverity": 2,
      "open": true,
      "resultState": "To Verify",
      "resultStatus": "New",
      "severity": "Low",
      "targets": Array [
        "test-project",
      ],
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1000990&projectid=328&pathid=1",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "checkmarx_finding:projectid=396&scanid=1001371&pathid=2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "Assigned To": "",
            "Column": "45",
            "Comment": "",
            "Custom": "",
            "DestColumn": "21",
            "DestFileName": "src/azure/utils.ts",
            "DestLine": "15",
            "DestName": "toLowerCase",
            "DestNodeId": "3",
            "FISMA 2014": "Identification And Authentication",
            "Line": "63",
            "Link": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=2",
            "NIST SP 800-53": "SC-4 Information in Shared Resources (P1)",
            "Name": "location",
            "NodeId": "1",
            "OWASP Mobile Top 10 2016": "",
            "OWASP Top 10 2013": "",
            "OWASP Top 10 2017": "A3-Sensitive Data Exposure",
            "PCI DSS v3.2": "",
            "Query": "Client Privacy Violation",
            "QueryPath": "JavaScript\\\\Cx\\\\JavaScript Medium Threat\\\\Client Privacy Violation Version:1",
            "Result Severity": "Medium",
            "Result State": "To Verify",
            "Result Status": "Recurrent",
            "SrcFileName": "src/steps/resource-manager/cosmosdb/converters.ts",
          },
        },
      ],
      "_type": "checkmarx_finding",
      "category": "sast-scan",
      "createdOn": undefined,
      "displayName": "Client Privacy Violation",
      "id": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=2",
      "name": "Client Privacy Violation",
      "numericSeverity": 3,
      "open": true,
      "resultState": "To Verify",
      "resultStatus": "Recurrent",
      "severity": "Medium",
      "targets": Array [
        "another-test-project",
      ],
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=2",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "checkmarx_finding:projectid=396&scanid=1001371&pathid=3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "Assigned To": "",
            "Column": "40",
            "Comment": "",
            "Custom": "",
            "DestColumn": "21",
            "DestFileName": "src/azure/utils.ts",
            "DestLine": "15",
            "DestName": "toLowerCase",
            "DestNodeId": "3",
            "FISMA 2014": "Identification And Authentication",
            "Line": "26",
            "Link": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=3",
            "NIST SP 800-53": "SC-4 Information in Shared Resources (P1)",
            "Name": "location",
            "NodeId": "1",
            "OWASP Mobile Top 10 2016": "",
            "OWASP Top 10 2013": "",
            "OWASP Top 10 2017": "A3-Sensitive Data Exposure",
            "PCI DSS v3.2": "",
            "Query": "Client Privacy Violation",
            "QueryPath": "JavaScript\\\\Cx\\\\JavaScript Medium Threat\\\\Client Privacy Violation Version:1",
            "Result Severity": "Medium",
            "Result State": "To Verify",
            "Result Status": "Recurrent",
            "SrcFileName": "src/steps/resource-manager/key-vault/converters.ts",
          },
        },
      ],
      "_type": "checkmarx_finding",
      "category": "sast-scan",
      "createdOn": undefined,
      "displayName": "Client Privacy Violation",
      "id": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=3",
      "name": "Client Privacy Violation",
      "numericSeverity": 3,
      "open": true,
      "resultState": "To Verify",
      "resultStatus": "Recurrent",
      "severity": "Medium",
      "targets": Array [
        "another-test-project",
      ],
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=3",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "checkmarx_finding:projectid=396&scanid=1001371&pathid=4",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "Assigned To": "",
            "Column": "40",
            "Comment": "",
            "Custom": "",
            "DestColumn": "21",
            "DestFileName": "src/azure/utils.ts",
            "DestLine": "15",
            "DestName": "toLowerCase",
            "DestNodeId": "3",
            "FISMA 2014": "Identification And Authentication",
            "Line": "31",
            "Link": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=4",
            "NIST SP 800-53": "SC-4 Information in Shared Resources (P1)",
            "Name": "location",
            "NodeId": "1",
            "OWASP Mobile Top 10 2016": "",
            "OWASP Top 10 2013": "",
            "OWASP Top 10 2017": "A3-Sensitive Data Exposure",
            "PCI DSS v3.2": "",
            "Query": "Client Privacy Violation",
            "QueryPath": "JavaScript\\\\Cx\\\\JavaScript Medium Threat\\\\Client Privacy Violation Version:1",
            "Result Severity": "Medium",
            "Result State": "To Verify",
            "Result Status": "Recurrent",
            "SrcFileName": "src/steps/resource-manager/cosmosdb/converters.ts",
          },
        },
      ],
      "_type": "checkmarx_finding",
      "category": "sast-scan",
      "createdOn": undefined,
      "displayName": "Client Privacy Violation",
      "id": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=4",
      "name": "Client Privacy Violation",
      "numericSeverity": 3,
      "open": true,
      "resultState": "To Verify",
      "resultStatus": "Recurrent",
      "severity": "Medium",
      "targets": Array [
        "another-test-project",
      ],
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=4",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "checkmarx_finding:projectid=396&scanid=1001371&pathid=1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "Assigned To": "",
            "Column": "25",
            "Comment": "",
            "Custom": "",
            "DestColumn": "25",
            "DestFileName": "src/steps/resource-manager/databases/mysql/client.test.ts",
            "DestLine": "129",
            "DestName": "json",
            "DestNodeId": "1",
            "FISMA 2014": "",
            "Line": "129",
            "Link": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=1",
            "NIST SP 800-53": "",
            "Name": "json",
            "NodeId": "1",
            "OWASP Mobile Top 10 2016": "",
            "OWASP Top 10 2013": "",
            "OWASP Top 10 2017": "",
            "PCI DSS v3.2": "",
            "Query": "Missing HSTS Header",
            "QueryPath": "JavaScript\\\\Cx\\\\JavaScript Medium Threat\\\\Missing HSTS Header Version:1",
            "Result Severity": "Medium",
            "Result State": "To Verify",
            "Result Status": "Recurrent",
            "SrcFileName": "src/steps/resource-manager/databases/mysql/client.test.ts",
          },
        },
      ],
      "_type": "checkmarx_finding",
      "category": "sast-scan",
      "createdOn": undefined,
      "displayName": "Missing HSTS Header",
      "id": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=1",
      "name": "Missing HSTS Header",
      "numericSeverity": 3,
      "open": true,
      "resultState": "To Verify",
      "resultStatus": "Recurrent",
      "severity": "Medium",
      "targets": Array [
        "another-test-project",
      ],
      "webLink": "https://cxprivatecloud.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1001371&projectid=396&pathid=1",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_account:cxprivatecloud",
      "_key": "checkmarx_account:cxprivatecloud|has|checkmarx_dast_scanner:Checkmarx DAST",
      "_toEntityKey": "checkmarx_dast_scanner:Checkmarx DAST",
      "_type": "checkmarx_account_has_dast_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_account:cxprivatecloud",
      "_key": "checkmarx_account:cxprivatecloud|has|checkmarx_team:109",
      "_toEntityKey": "checkmarx_team:109",
      "_type": "checkmarx_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_team:109",
      "_key": "checkmarx_team:109|has|checkmarx_project:328",
      "_toEntityKey": "checkmarx_project:328",
      "_type": "checkmarx_team_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_team:109",
      "_key": "checkmarx_team:109|has|checkmarx_project:352",
      "_toEntityKey": "checkmarx_project:352",
      "_type": "checkmarx_team_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_team:109",
      "_key": "checkmarx_team:109|has|checkmarx_project:396",
      "_toEntityKey": "checkmarx_project:396",
      "_type": "checkmarx_team_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_project:328",
      "_key": "checkmarx_project:328|has|checkmarx_scan:1000990",
      "_toEntityKey": "checkmarx_scan:1000990",
      "_type": "checkmarx_project_has_scan",
      "displayName": "HAS",
    },
    Object {
      "_class": "PERFORMED",
      "_fromEntityKey": "checkmarx_dast_scanner:Checkmarx DAST",
      "_key": "checkmarx_dast_scanner:Checkmarx DAST|performed|checkmarx_scan:1000990",
      "_toEntityKey": "checkmarx_scan:1000990",
      "_type": "checkmarx_dast_scanner_performed_scan",
      "displayName": "PERFORMED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "checkmarx_project:396",
      "_key": "checkmarx_project:396|has|checkmarx_scan:1001371",
      "_toEntityKey": "checkmarx_scan:1001371",
      "_type": "checkmarx_project_has_scan",
      "displayName": "HAS",
    },
    Object {
      "_class": "PERFORMED",
      "_fromEntityKey": "checkmarx_dast_scanner:Checkmarx DAST",
      "_key": "checkmarx_dast_scanner:Checkmarx DAST|performed|checkmarx_scan:1001371",
      "_toEntityKey": "checkmarx_scan:1001371",
      "_type": "checkmarx_dast_scanner_performed_scan",
      "displayName": "PERFORMED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "checkmarx_scan:1000990",
      "_key": "checkmarx_scan:1000990|identified|checkmarx_finding:projectid=328&scanid=1000990&pathid=1",
      "_toEntityKey": "checkmarx_finding:projectid=328&scanid=1000990&pathid=1",
      "_type": "checkmarx_scan_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "checkmarx_scan:1001371",
      "_key": "checkmarx_scan:1001371|identified|checkmarx_finding:projectid=396&scanid=1001371&pathid=2",
      "_toEntityKey": "checkmarx_finding:projectid=396&scanid=1001371&pathid=2",
      "_type": "checkmarx_scan_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "checkmarx_scan:1001371",
      "_key": "checkmarx_scan:1001371|identified|checkmarx_finding:projectid=396&scanid=1001371&pathid=3",
      "_toEntityKey": "checkmarx_finding:projectid=396&scanid=1001371&pathid=3",
      "_type": "checkmarx_scan_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "checkmarx_scan:1001371",
      "_key": "checkmarx_scan:1001371|identified|checkmarx_finding:projectid=396&scanid=1001371&pathid=4",
      "_toEntityKey": "checkmarx_finding:projectid=396&scanid=1001371&pathid=4",
      "_type": "checkmarx_scan_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "checkmarx_scan:1001371",
      "_key": "checkmarx_scan:1001371|identified|checkmarx_finding:projectid=396&scanid=1001371&pathid=1",
      "_toEntityKey": "checkmarx_finding:projectid=396&scanid=1001371&pathid=1",
      "_type": "checkmarx_scan_identified_finding",
      "displayName": "IDENTIFIED",
    },
  ],
  "encounteredTypes": Array [
    "checkmarx_account",
    "checkmarx_dast_scanner",
    "checkmarx_account_has_dast_scanner",
    "checkmarx_team",
    "checkmarx_account_has_team",
    "checkmarx_project",
    "checkmarx_team_has_project",
    "checkmarx_scan",
    "checkmarx_project_has_scan",
    "checkmarx_dast_scanner_performed_scan",
    "checkmarx_finding",
    "checkmarx_scan_identified_finding",
  ],
  "numCollectedEntities": 13,
  "numCollectedRelationships": 14,
}
`;
